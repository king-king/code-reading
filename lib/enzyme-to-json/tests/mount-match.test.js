/* eslint-env jest */

import React from 'react';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import mountToJson from '../src/mount';
import {Table, TobeList} from './fixtures/match-object';

Enzyme.configure({adapter: new Adapter()});

const deepOptions = {mode: 'deep'};
const shallowOptions = {mode: 'shallow'};

describe('Table', () => {
  // Mock child components to omit irrelevant props for shallow match,
  // especially event handlers so not to be redundant with interaction tests.
  function addRow() {}
  function deleteRow() {}

  // suggestions in a game of Cluedo
  const fields = [
    {key: 'who', label: 'qui'},
    {key: 'where', label: 'où'},
    {key: 'which', label: 'lequel'},
  ];
  const records = [
    {id: 3, who: 'Plum', where: 'study', which: 'spanner'},
    {id: 2, who: 'Peacock', where: 'library', which: 'rope'},
    {id: 1, who: 'Green', where: 'conservatory', which: 'pipe'},
  ];

  describe('matches class components to maximum depth', () => {
    const $it = mount(
      <Table
        addRow={addRow}
        deleteRow={deleteRow}
        fields={fields}
        records={records}
      />,
    );

    test('at thead', () => {
      const received = mountToJson($it.find('thead'), deepOptions);
      expect(received).toMatchSnapshot();
    });

    test('at tr', () => {
      // Be careful. Don’t duplicate the application code.
      const i = 1;
      const received = mountToJson($it.find('tbody tr').at(i), deepOptions);
      expect(received).toMatchSnapshot();
    });
  });

  describe('matches class components to minimum depth', () => {
    const $it = mount(
      <Table
        addRow={addRow}
        deleteRow={deleteRow}
        fields={fields}
        records={records}
      />,
    );

    test('at table', () => {
      const received = mountToJson($it.find('table'), shallowOptions);
      expect(received).toMatchSnapshot();
    });

    test('at thead', () => {
      const received = mountToJson($it.find('thead'), shallowOptions);
      expect(received).toMatchSnapshot();
    });

    test('at tbody', () => {
      const received = mountToJson($it.find('tbody'), shallowOptions);
      expect(received).toMatchSnapshot();
    });

    test('at tr', () => {
      // Be careful. Don’t duplicate the application code.
      const i = 1;
      const received = mountToJson($it.find('tbody tr').at(i), shallowOptions);
      expect(received).toMatchSnapshot();
    });
  });
});

describe('TobeList', () => {
  // statements which are to be, or not to be :)
  const items = [
    {
      id: 0,
      is: false,
      text: 'testing is painless',
    },
    {
      id: 1,
      is: true,
      text: 'more is less',
    },
    {
      id: 2,
      is: true,
      text: 'less is more',
    },
  ];

  it('matches functional components to maximum depth', () => {
    const received = mountToJson(
      mount(<TobeList items={items} />),
      deepOptions,
    );

    expect(received).toMatchSnapshot();
  });

  it('matches functional components to minimum depth at a descendant', () => {
    const received = mountToJson(mount(<TobeList items={items} />).find('ul'), {
      mode: 'shallow',
      noKey: true,
    });

    expect(received).toMatchSnapshot();
  });

  it('matches functional components to either depth at a descendant', () => {
    const element = <TobeList items={items} />;
    const receivedDeep = mountToJson(
      mount(element)
        .find('li')
        .at(1),
      deepOptions,
    );
    const receivedShallow = mountToJson(
      mount(element)
        .find('li')
        .at(1),
      shallowOptions,
    );

    // When all descendants are DOM nodes, deep and shallow are equivalent.
    expect(receivedDeep).toMatchSnapshot();
    expect(receivedShallow).toMatchSnapshot();
  });
});
