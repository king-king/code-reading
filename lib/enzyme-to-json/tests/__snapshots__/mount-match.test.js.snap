// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table matches class components to maximum depth at thead 1`] = `
<thead>
  <tr>
    <th>
      <button
        onClick={[Function]}
        title="add row"
      >
        +
      </button>
    </th>
    <th
      key="who"
      scope="col"
    >
      qui
    </th>
    <th
      key="where"
      scope="col"
    >
      où
    </th>
    <th
      key="which"
      scope="col"
    >
      lequel
    </th>
  </tr>
</thead>
`;

exports[`Table matches class components to maximum depth at tr 1`] = `
<tr>
  <td>
    <button
      onClick={[Function]}
      title="delete row"
    >
      -
    </button>
  </td>
  <td
    key="who"
  >
    Peacock
  </td>
  <td
    key="where"
  >
    library
  </td>
  <td
    key="which"
  >
    rope
  </td>
</tr>
`;

exports[`Table matches class components to minimum depth at table 1`] = `
<table>
  <TableHead
    addRow={[Function]}
    fields={
      Array [
        Object {
          "key": "who",
          "label": "qui",
        },
        Object {
          "key": "where",
          "label": "où",
        },
        Object {
          "key": "which",
          "label": "lequel",
        },
      ]
    }
  />
  <tbody>
    <TableRow
      deleteRow={[Function]}
      fields={
        Array [
          Object {
            "key": "who",
            "label": "qui",
          },
          Object {
            "key": "where",
            "label": "où",
          },
          Object {
            "key": "which",
            "label": "lequel",
          },
        ]
      }
      key="3"
      record={
        Object {
          "id": 3,
          "where": "study",
          "which": "spanner",
          "who": "Plum",
        }
      }
    />
    <TableRow
      deleteRow={[Function]}
      fields={
        Array [
          Object {
            "key": "who",
            "label": "qui",
          },
          Object {
            "key": "where",
            "label": "où",
          },
          Object {
            "key": "which",
            "label": "lequel",
          },
        ]
      }
      key="2"
      record={
        Object {
          "id": 2,
          "where": "library",
          "which": "rope",
          "who": "Peacock",
        }
      }
    />
    <TableRow
      deleteRow={[Function]}
      fields={
        Array [
          Object {
            "key": "who",
            "label": "qui",
          },
          Object {
            "key": "where",
            "label": "où",
          },
          Object {
            "key": "which",
            "label": "lequel",
          },
        ]
      }
      key="1"
      record={
        Object {
          "id": 1,
          "where": "conservatory",
          "which": "pipe",
          "who": "Green",
        }
      }
    />
  </tbody>
</table>
`;

exports[`Table matches class components to minimum depth at tbody 1`] = `
<tbody>
  <TableRow
    deleteRow={[Function]}
    fields={
      Array [
        Object {
          "key": "who",
          "label": "qui",
        },
        Object {
          "key": "where",
          "label": "où",
        },
        Object {
          "key": "which",
          "label": "lequel",
        },
      ]
    }
    key="3"
    record={
      Object {
        "id": 3,
        "where": "study",
        "which": "spanner",
        "who": "Plum",
      }
    }
  />
  <TableRow
    deleteRow={[Function]}
    fields={
      Array [
        Object {
          "key": "who",
          "label": "qui",
        },
        Object {
          "key": "where",
          "label": "où",
        },
        Object {
          "key": "which",
          "label": "lequel",
        },
      ]
    }
    key="2"
    record={
      Object {
        "id": 2,
        "where": "library",
        "which": "rope",
        "who": "Peacock",
      }
    }
  />
  <TableRow
    deleteRow={[Function]}
    fields={
      Array [
        Object {
          "key": "who",
          "label": "qui",
        },
        Object {
          "key": "where",
          "label": "où",
        },
        Object {
          "key": "which",
          "label": "lequel",
        },
      ]
    }
    key="1"
    record={
      Object {
        "id": 1,
        "where": "conservatory",
        "which": "pipe",
        "who": "Green",
      }
    }
  />
</tbody>
`;

exports[`Table matches class components to minimum depth at thead 1`] = `
<thead>
  <tr>
    <th>
      <Button
        onClick={[Function]}
        text="+"
        title="add row"
      />
    </th>
    <th
      key="who"
      scope="col"
    >
      qui
    </th>
    <th
      key="where"
      scope="col"
    >
      où
    </th>
    <th
      key="which"
      scope="col"
    >
      lequel
    </th>
  </tr>
</thead>
`;

exports[`Table matches class components to minimum depth at tr 1`] = `
<tr>
  <td>
    <Button
      onClick={[Function]}
      text="-"
      title="delete row"
    />
  </td>
  <td
    key="who"
  >
    Peacock
  </td>
  <td
    key="where"
  >
    library
  </td>
  <td
    key="which"
  >
    rope
  </td>
</tr>
`;

exports[`TobeList matches functional components to either depth at a descendant 1`] = `
<li
  style={
    Object {
      "textDecoration": "none",
    }
  }
>
  more is less
</li>
`;

exports[`TobeList matches functional components to either depth at a descendant 2`] = `
<li
  style={
    Object {
      "textDecoration": "none",
    }
  }
>
  more is less
</li>
`;

exports[`TobeList matches functional components to maximum depth 1`] = `
<ul>
  <li
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
  >
    testing is painless
  </li>
  <li
    style={
      Object {
        "textDecoration": "none",
      }
    }
  >
    more is less
  </li>
  <li
    style={
      Object {
        "textDecoration": "none",
      }
    }
  >
    less is more
  </li>
</ul>
`;

exports[`TobeList matches functional components to minimum depth at a descendant 1`] = `
<ul>
  <TobeItem
    is={false}
    text="testing is painless"
  />
  <TobeItem
    is={true}
    text="more is less"
  />
  <TobeItem
    is={true}
    text="less is more"
  />
</ul>
`;
